### ISSUE TITLE
Error: Both secret and user tokens are not set. Either client.connectUser wasn't called or client.disconnect was called

### ISSUE THREAD

**Issue by moroshko** [GetStream/stream-chat-react#1822]
I follow the example [here](https://getstream.io/chat/docs/react/instantiating_the_client/#examples) to instantiate the client, and get the following error when rendering a `ChannelList`:

> Both secret and user tokens are not set. Either client.connectUser wasn't called or client.disconnect was called

<img width="457" alt="Screen Shot 2022-10-20 at 3 12 56 pm" src="https://user-images.githubusercontent.com/259753/196855107-c5bbc838-b12c-4125-9cc0-8f75e5f9a341.png">

I use:

```
react: 18.2.0
stream-chat: 8.0.0
stream-chat-react: 10.3.1
```

What am I doing wrong?

[CodeSandbox](https://codesandbox.io/s/pedantic-yalow-vqp0tl?file=/src/App.tsx) (Make sure to edit the variables at the top to your own!)

```typescript
const chatKey = "...";
const userId = "...";
const chatToken = "...";

const client = StreamChat.getInstance(chatKey);

export default function App() {
  const [chatClient, setChatClient] = useState<StreamChat | null>(null);

  useEffect(() => {
    const initChat = async () => {
      await client.connectUser({ id: userId }, chatToken);
      setChatClient(client);
    };

    initChat();

    return () => {
      setChatClient(null);
      client.disconnectUser();
    };
  }, []);

  if (chatClient === null) {
    return null;
  }

  return (
    <Chat client={chatClient}>
      <ChannelList
        filters={{
          type: "messaging",
          members: { $in: [userId] }
        }}
      />
    </Chat>
  );
}
```

**Comment by petyosi** [comment]
The chat client needs to be instantiated in the hook itself - [check this out](https://codesandbox.io/s/goofy-rain-o5mzuh?file=/src/App.tsx).

**Comment by moroshko** [comment]
@petyosi Thanks for this!
1. I didn't realize the importance of calling `StreamChat.getInstance()` from `useEffect()`. Now, I'm curious to understand why this makes a difference. Also, if it's not clearly documented yet (I didn't see it in the docs), I think it worth emphasizing that!
2. In the example you shared above, there is a warning: "Consecutive calls to connectUser is detected, ideally you should only call this function once in your app.". I assume it's because React 18 calls `useEffect` twice in dev mode. How would you recommend avoiding this warning?

**Comment by petyosi** [comment]
Both questions are related to the same problem actually, `useEffect` and its "destructor" being called twice. If you notice, I changed the destructor in the sample I sent as well. 

We are looking into shipping a ready-made hook for integrators to avoid mistakes like this (and potentially fix the dev mode), but we have not done this yet.

**Comment by moroshko** [comment]
There is also a warning in your example about the dependency array of `useEffect` that's missing the `chatClient`. And, I'm afraid if we add the `chatClient`, it could result in even more calls to `connectUser`.
What's the recommended approach for now @petyosi to avoid all these warnings?

**Comment by petyosi** [comment]
That's tricky as I mentioned. FWIW, this is the work in progress in [this branch](https://github.com/GetStream/stream-chat-react/compare/poc/use-client) for the hook we want to bundle.

**Comment by moroshko** [comment]
@petyosi I took your example, and tried to make one step forward by adding the ability to switch users. However, when I switch to User 2, I get this error:
> Use client.disconnect() before trying to connect as a different user. connectUser was called twice.

Any ideas?

[CodeSandbox](https://codesandbox.io/s/nervous-bhaskara-3lxqlk?file=/src/constants.ts:0-194)

**Comment by petyosi** [comment]
You can try this version of the hook - https://github.com/GetStream/website-react-examples/blob/master/social-messenger-ts/src/hooks/useConnectUser.ts

**Comment by moroshko** [comment]
@petyosi This `useConnectUser` hook doesn't work, unfortunately. 

As you can see, since `useEffect` is being called twice by React, `setChatClient(client)` is being called, but then immediately after, `useEffect`'s cleanup function is run causing  `setChatClient(null)`. This results in the `chatClient` being `null` for the consumer so the chat never renders.

[CodeSandbox](https://codesandbox.io/s/modest-violet-wxoh5h?file=/src/useConnectUser.ts)

Can we please reopen this issue until there is a way forward?

@arnautov-anton Any ideas? Feels somewhat related to #1487.

**Comment by petyosi** [comment]
@moroshko you are correct about #1487, this is the same problem. We're discussing the problem internally with @arnautov-anton and the rest of the team. Like I've mentioned, we don't have a ready-to-ship solution. This issue is a support ticket where we're trying to find a reasonable solution that works for you. 

Back to the sandbox created, as far as I can see, you modified the hook I referred to. [Here's how it's supposed to be ](https://codesandbox.io/s/musing-jasper-ryssg1?file=/src/App.tsx) integrated. Let me know if this works for you.

**Comment by moroshko** [comment]
Thanks for this @petyosi. I indeed modified the hook a bit without realizing that `new StreamChat(apiKey)` and `StreamChat.getInstance(apiKey)` are very different. If I use `new StreamChat(apiKey)` in my sandbox above, it works.

When I read the [docs](https://getstream.io/chat/docs/react/?language=javascript#clients-and-users):
```typescript
const client = StreamChat.﻿getInstance﻿(﻿﻿"YOUR_API_KEY")﻿;﻿
// you can still use new StreamChat("api_key");﻿
```
I got the impression that `StreamChat.getInstance(apiKey)` is the new way to get an instance, and that it's the recommended way. Isn't it the case? By the looks of it, the two methods produce very different behavior. Is this documented somewhere?

Also, in your sandbox, if you switch to User 2 and then back to User 1, the following error is logged:
> StreamChat error code 70: QueryChannels failed with error: "1 channels match your query but cannot be returned because you don't have access to them. Did you forget to include {members: $in: ["fragrant-river-1"]}?

Can it be avoided?

**Comment by petyosi** [comment]
@moroshko the singleton vs non-singleton usage is not documented. The hook implementation is meant to be internal, so integrators should not deal with those matters. 

regarding your next question: pretty much what the error message suggests to do should solve the problem. 

Given that the original topic of this issue has been addressed. I may suggest you to contact support@getstream.io with further questions - our support engineers will assist you accordingly.

**Comment by moroshko** [comment]
> pretty much what the error message suggests to do should solve the problem.

Not sure I'm following. `{ members: { $in: [ ... ] } }` is provided. I'll create a separate issue.
